modules/common/instance/instance.go

```diff
@@ -0,0 +1,85 @@
+/*
+Copyright 2020 Red Hat
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package instance
+
+import (
+	"context"
+
+	helper "github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+	util "github.com/openstack-k8s-operators/lib-common/modules/common/util"
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// PatchInstance - Patch an instance's metadata and/or status if they have changed (based
+// on comparison with its old state as represented in the helper.Helper)
+//
+// NOTE: This function is mainly intended for use in Podified Operators with their
+// deferred-instance-persistence pattern within the reconcile loop.  If used in this manner,
+// any error returned by this function should be set as the error returned by the encompassing
+// Reconcile(...) function.
+//
+// Example:
+//
+// func (r *SomeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, _err error) {
+//     ...
+//     defer func() {
+//         ...
+//         err := instance.PatchInstance(ctx, h, instance)
+//         if err != nil {
+//             _err = err
+//             return
+//         }
+//     }
+//     ...
+// }
+//
+func PatchInstance(ctx context.Context, h *helper.Helper, instance client.Object) error {
+	var err error
+
+	if err = h.SetAfter(instance); err != nil {
+		util.LogErrorForObject(h, err, "Set after and calc patch/diff", instance)
+		return err
+	}
+
+	changes := h.GetChanges()
+	patch := client.MergeFrom(h.GetBeforeObject())
```

Comment:
I understand this is just a refactoring (moving the code around), but wouldn't it make sense to adjust it a bit by avoiding that patch calculated twice - here, and when we do calculate changes inside of SetAfter https://github.com/abays/lib-common/blob/feb92dd06374e5e2591ef53780b8fe788c60ca46/modules/common/helper/helper.go#L147 ?

we can store patch in the helper and just pick it here. That would also help to avoid conflicts when forming patches at different time moments, so that the most recent one could be bease on the no longer actual GetChanges result?  And the problems with applying it, would shift to the end of the patching, which looks more precise...

Reply:
`client.MergeFrom()`  only defines the _base_ of the patch, but does not do the actual diffing. The second diff calculation actually happens inside the `h.GetClient().Patch()` call. As that is out of our control I'm not sure how can we reduce the number of patch calculations.

Reply:
Thank you for explaining. Perhaps we still could avoid doing a 2nd call for ``client.MergeFrom()``

Reply:
I dont see a second call to `h.GetChanges` i only see one here.

Reply:
L199 should result in `h.changes` being set and then we acquire those cached changes at L204.  So it seems like we already are using the stored result.  Perhaps I am misunderstanding what you mean.

Reply:
I was confused by client.MergeFrom(), sorry